### PR TITLE
Hex Tiling

### PR DESCRIPTION
**Pull request by RuffledPlume**
Base Sampling |  Hex Tiling Sampling
:-------------------------:|:-------------------------:
image | image

### CODE REVIEW

File: schemas/materials.schema.json [hunk 1]
```diff
@@ -87,6 +87,23 @@
         "type": "number",
         "description": "The gloss of specular reflections. Defaults to 0. Commonly ranges from 10 to 500."
       },
+      "hexTilingScale": {
+        "type": "number",
+        "description": "The scale of the hexagonal tiling pattern. Defaults to 0.0, i.e. no tiling."
```
Comment (aHooder) [inline schemas/materials.schema.json]: It probably makes sense for this to be its own field & not directly tied to `textureScale`, but the description could be clearer about acting like both the toggle & scale of hex tiling.

File: schemas/materials.schema.json [hunk 2]
```diff
@@ -87,6 +87,23 @@
...
+      },
+      "hexTilingBlend": {
+        "type": "number",
+        "description": "The blending of the hexagonal tiling pattern. Defaults to 4.0."
+      },
+      "hexTilingMode": {
+        "type": "string",
+        "description": "",
```
Comment (aHooder) [inline schemas/materials.schema.json]: What do these modes do?

File: src/main/resources/rs117/hd/scene_frag.glsl [hunk 3]
```diff
@@ -205,6 +206,16 @@ void main() {
             fragPos += TBN * fragDelta;
         #endif
 
+        // Build HexData, if Terrain use world space XZ otherwise UVs
+        HexData hex1 = buildHexData(uv1, IN.position / TILE_SIZE, material1.hexTilingScale, material1.hexTilingBlend, getMaterialHexTilingMode(material1));
```
Comment (aHooder) [inline src/main/resources/rs117/hd/scene_frag.glsl]: Does this hex data building for every material cost anything, or is it basically the same as not using it or gating it behind a branch? Some quick & dirty measurements of the impact would be nice 🙂

File: schemas/materials.schema.json (hunk 1 above)
Comment (aHooder) [reply schemas/materials.schema.json]: The scale should also be changed to behave similarly to `textureScale`, in that larger numbers mean larger appearance of the texture, since that way around is just more intuitive. The `textureScale` stuff does this by calculating the reciprocal during normalization if I remember correctly, or it might be during material UBO filling.

File: src/main/resources/rs117/hd/utils/hex_tiling.glsl
Comment (aHooder) [inline src/main/resources/rs117/hd/utils/hex_tiling.glsl:1]: Does this still need testing on macOS? 👀

**Review by aHooder — CHANGES_REQUESTED** [review]
Not much to change, but I think it would make sense to make a hex tiling bool to tightly packed as a flag in the material data UBO, and repurpose the otherwise unused texture scale floats instead of adding new fields. Also some comments & general questions 🙂

File: schemas/materials.schema.json (hunk 2 above)
Comment (RuffledPlume) [reply schemas/materials.schema.json]: **OFFSET**
Will randomly offset & scale the base UVs of the Hexagon only

**OFFSET_WITH_MIRROR**
Flips the U and/or V components of the Hexagon

**OFFSET_WITH_ROTATION**
Applies a Discrete rotation to the UV


Offset is the cheapest/basic mode, with Mirror being a cheaper version of rotation since mirror works by applying a step to the hex hash which is used to calculate the offset. Whereas the rotation calls sin/cos to apply the rotation

File: src/main/resources/rs117/hd/scene_frag.glsl (hunk 3 above)
Comment (RuffledPlume) [reply src/main/resources/rs117/hd/scene_frag.glsl]: Its gated behind a branch at the moment, materials default `hexTilingBlend` to 0. Which will early out within: 
```glsl
HexData buildHexData(vec2 uv, vec3 fragPos, float scale, float blend, int mode) {
    HexData h;
    if (scale <= 0.0) {
        h.enabled = false;
        return h;
    }
    h.enabled = true;
```

File: schemas/materials.schema.json (hunk 1 above)
Comment (RuffledPlume) [reply schemas/materials.schema.json]: Didn't touch the materialData flags since I'd like to move allot of those bits over to the ModelData struct once that is merged, also it kinda means there'd be two ways to disable hex tiling since a scale of 0 would break the sampling mode.

Edit: Ah you mean the Material Struct flags not the ModelOverride. That could work. but we'd still have the fact that scaling could break near 0

**Comment by 3-X** [comment]
Tested on MacOS this works without any issues
image

File: schemas/materials.schema.json (hunk 1 above)
Comment (aHooder) [reply schemas/materials.schema.json]: Yeah, I should've specified. I was thinking we can keep `hexTilingScale` just as it is, default to zero, then whenever it's set to something non-zero, consider hex tiling as enabled, and set a flag in the existing `flags` in `UBOMaterials`. Then, since the shader now knows when to use hex tiling, we can repurpose the existing first float component of the `textureScale`, so we don't need to add a whole new `hexTilingScale` field to the material struct. The second component could be used for the `hexTilingBlend`. I suppose it's not really necessary, but I'm not too sure how far off the UBO size limit we currently are with materials either. Probably worth changing imo :eyes:

File: schemas/materials.schema.json (hunk 2 above)
Comment (aHooder) [reply schemas/materials.schema.json]: Hmm, so these are kind of like "normal", "slightly more random appearance" and "even more random appearance"? Since the rotation is random, and the first mode doesn't rotate, the second rotates but only cheaply, and the third allows more arbitrary rotations?

Comment (RuffledPlume) [reply schemas/materials.schema.json]: Yup, it depends on where its being used. On smaller surfaces offset/scale is usually enough, but on terrain rotation brakes up repeated patterns even more

Comment (aHooder) [reply schemas/materials.schema.json]: Might be worth renaming these since it's a bit confusing, but I'm not sure what to exactly haha. Will have a think about it :slightly_smiling_face:

File: src/main/java/rs117/hd/config/HexTilingMode.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package rs117.hd.config;
+
+public enum HexTilingMode {
+	OFFSET,
+	OFFSET_WITH_MIRROR,
+	OFFSET_WITH_ROTATION
```
Comment (aHooder) [inline src/main/java/rs117/hd/config/HexTilingMode.java:6]: The first 2 modes seem to work, but applying `OFFSET_WITH_ROTATION` to tiles reveals the UV tiling:
Image
This is with https://www.oxpal.com/uv-checker-texture.html saved as `gravel.png` in the texture folder.

This doesn't seem to be easily fixable, and since the other two modes work well we could just remove this mode for now, if you prefer.

File: src/main/java/rs117/hd/opengl/uniforms/UBOMaterials.java [hunk 5]
```diff
@@ -24,6 +24,8 @@ public static class MaterialStruct extends StructProperty {
 		public Property displacementScale = addProperty(PropertyType.Float, "displacementScale");
 		public Property specularStrength = addProperty(PropertyType.Float, "specularStrength");
 		public Property specularGloss = addProperty(PropertyType.Float, "specularGloss");
+		public Property hexTilingScale = addProperty(PropertyType.Float, "hexTilingScale"); // TODO: combine with textureScale if sensible
+		public Property hexTilingBlend = addProperty(PropertyType.Float, "hexTilingBlend"); //       & reuse unused textureScale component
```
Comment (aHooder) [inline src/main/java/rs117/hd/opengl/uniforms/UBOMaterials.java:28]: I've added a couple of TODOs here, since it _feels_ sensible to me that the scale of the underlying texture could/should be tide to the hex tiling scale. This would let us also reuse the `textureScale` material components for both of these, so we don't need to add a few kBs to the material UBO. If you prefer, we can just keep these TODOs around and merge the PR anyway though.

**Review by aHooder — CHANGES_REQUESTED** [review]
Just one or two issues remaining. Regarding these `LAYERED_SAMPLE` macros, is the idea simply to check if materials are identical, and only sample when they're not identical, to help out any compilers which might struggle with this?
